### PR TITLE
JAVA-3129: Make all event listener methods default methods

### PR DESCRIPTION
**Pull request by jstewart-mongo**
Evergreen patch: https://evergreen.mongodb.com/version/5d5db4e93627e025a533ba24

### CODE REVIEW

**Comment by jstewart-mongo** [comment]
https://docs.oracle.com/javase/specs/jls/se8/html/jls-13.html#jls-13.5.6 describes the effects of changing an abstract method to default.

File: driver-core/src/main/com/mongodb/event/ClusterListenerAdapter.java [hunk 1]
```diff
@@ -22,16 +22,5 @@
...
-    }
-
-    @Override
-    public void clusterClosed(final ClusterClosedEvent event) {
-    }
-
-    @Override
-    public void clusterDescriptionChanged(final ClusterDescriptionChangedEvent event) {
-    }
+public class ClusterListenerAdapter implements ClusterListener {
```
[the remark this answers is not in the document]
Comment (jstewart-mongo) [reply driver-core/src/main/com/mongodb/event/ClusterListenerAdapter.java]: Done

File: driver-core/src/main/com/mongodb/event/ConnectionPoolListenerAdapter.java [hunk 2]
```diff
@@ -22,37 +22,5 @@
...
-    }
-
-    @Override
-    public void connectionAdded(final ConnectionAddedEvent event) {
-    }
-
-    @Override
-    public void connectionRemoved(final ConnectionRemovedEvent event) {
-    }
+public class ConnectionPoolListenerAdapter implements ConnectionPoolListener {
```
[the remark this answers is not in the document]
Comment (jstewart-mongo) [reply driver-core/src/main/com/mongodb/event/ConnectionPoolListenerAdapter.java]: Done

File: driver-core/src/main/com/mongodb/event/ServerListenerAdapter.java [hunk 3]
```diff
@@ -22,16 +22,5 @@
...
-    }
-
-    @Override
-    public void serverClosed(final ServerClosedEvent event) {
-    }
-
-    @Override
-    public void serverDescriptionChanged(final ServerDescriptionChangedEvent event) {
-    }
+public class ServerListenerAdapter implements ServerListener {
```
[the remark this answers is not in the document]
Comment (jstewart-mongo) [reply driver-core/src/main/com/mongodb/event/ServerListenerAdapter.java]: Done

File: driver-core/src/main/com/mongodb/event/ServerMonitorListenerAdapter.java [hunk 4]
```diff
@@ -22,16 +22,5 @@
...
-    }
-
-    @Override
-    public void serverHeartbeatSucceeded(final ServerHeartbeatSucceededEvent event) {
-    }
-
-    @Override
-    public void serverHeartbeatFailed(final ServerHeartbeatFailedEvent event) {
-    }
+public class ServerMonitorListenerAdapter implements ServerMonitorListener {
```
[the remark this answers is not in the document]
Comment (jstewart-mongo) [reply driver-core/src/main/com/mongodb/event/ServerMonitorListenerAdapter.java]: Done

**Comment by jstewart-mongo** [comment]
I have performed a very simple test to check for binary compatibility between Java 7 and Java 8 when changing abstract methods to default methods in an interface. For the Java 7 test, I created 3 files:
### MyInterface.java
``` java
public interface MyInterface {
    void doSomething();
}
```
### MyAdapter.java
``` java
public class MyAdapter implements MyInterface {
    public void doSomething() {
        System.out.println("ran test");
    }
}
```
### AdapterTest.java
``` java
public class AdapterTest {
    public static void main(String[] args) {
        MyAdapter adapter = new MyAdapter();
        adapter.doSomething();
    }
}
```
I compiled MyAdapter.java and created a JAR file from the resulting classes:
``` shell
# javac MyAdapter.java
# jar cvf classes7.jar *.class
# rm -f My*.class
```
Next, I compiled AdapterTest.java and executed it:
``` shell
# javac -cp ".:./classes7.jar" AdapterTest.java 
# java -cp ".:./classes7.jar" AdapterTest
ran test
```
Next, I converted the interface to use a default method for Java 8 and updated MyAdapter.java accordingly:
### MyInterface.java
``` java
public interface MyInterface {
    default void doSomething() {
        System.out.println("ran test");
    }
}
```
### MyAdapter.java
``` java
public class MyAdapter implements MyInterface {
}
```
I compiled MyAdapter.java using Java 8 and created another JAR file from the resulting classes:
``` shell
# javac MyAdapter.java 
# jar cvf classes8.jar MyAdapter.class MyInterface.class
# rm -f My*.class
```
Finally, I reran AdapterTest with the new JAR file:
``` shell
# java -cp ".:./classes8.jar" AdapterTest
ran test
```
I ran other tests where the printed string was changed in the default method and that ran without error. I was also able to add default methods to the interface without error. As expected, changing the signature of the default method `doSomething` generated a `NoSuchMethodError`.

File: driver-core/src/main/com/mongodb/event/ClusterListenerAdapter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/event/ClusterListenerAdapter.java]: Does this class need any methods now ClusterListener has all the defaults?

**Comment by jyemin** [comment]
It's not ideal to start with deprecated classes.  We could deprecate in 3.12 and remove in 4.0, but if anyone's actually using them it would be annoying to remove use of the adapters in 3.x because then they'd have to add empty implementations of all the methods, and only after upgrading to 4.x could they then remove them.

File: driver-core/src/main/com/mongodb/event/ClusterListenerAdapter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jstewart-mongo) [reply driver-core/src/main/com/mongodb/event/ClusterListenerAdapter.java]: No, and I have removed all the methods from this class.

**Comment by jstewart-mongo** [comment]
Since I am merging this fix into the 4.x branch, and the goal was to remove all deprecated elements, we should not add more with this fix.

**Comment by rozza** [comment]
Personally I'd be happy with the default implementations (Adapters) not being deprecated at all - but I can see why they aren't really needed, so happy either way.

**Pull request closed without merging**